Fix/bogo handle large product

kzamanbd opened the pull request:
Simplifies frontend logic by using product info objects directly instead of mapping by product IDs. Removes unused product list data from PHP scripts and JS, and updates REST API to include currency in product info. Improves code maintainability and reduces unnecessary data fetching.

* Closes https://github.com/getdokan/storegrowth-sales-booster-pro/issues/181

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Dynamic, searchable product selector for BOGO settings with real-time filtering.
  * API responses now include enriched product info (name, price, currency) for improved displays.
  * Vendor dashboards now filter to show the current vendor's products.
  * Select inputs now accept extra attributes/handlers passed from parent components.

* **Bug Fixes**
  * Prevents errors when offer data is missing and improves product display logic.

* **Chores**
  * Removed debug logging and tightened equality checks.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kzamanbd committed:
Add Product REST controller and service provider

Introduces ProductController for handling product-related REST API endpoints and registers it via a new ServiceProvider class. Updates Bootstrap to add the core service provider to the container. Also removes a debug console.log from the restApi.js utility.

kzamanbd committed:
Enhance BOGO product selection and info handling

Refactored BasicInfo component to fetch product lists via API and update product info fields dynamically. Improved SelectBox to accept additional props, added debounced product search, and updated field value handling for offered and different products. Updated BogoController to include detailed product info in REST responses. Fixed strict equality checks in CreateBogo and improved code consistency.

kzamanbd committed:
Refactor BOGO product info handling and API response

Simplifies frontend logic by using product info objects directly instead of mapping by product IDs. Removes unused product list data from PHP scripts and JS, and updates REST API to include currency in product info. Improves code maintainability and reduces unnecessary data fetching.

kzamanbd committed:
Merge branch 'develop' into fix/bogo-handle-large-product

mrabbani commented on includes/REST/ProductController.php:
```diff
@@ -0,0 +1,305 @@
+            );
+        }
+
+		return $args;
```
Apply filters for Dokan integration to intercept the author param for vendors.

mrabbani commented on includes/REST/ProductController.php:
```diff
@@ -0,0 +1,305 @@
+		// author id
+		if( ! current_user_can( 'manage_options' ) ) {
+			$args['author'] = ! isset( $request['id'] ) ? get_current_user_id() : $request['id'];
+		}
```
Pls move this section to Dokan Integration.

mrabbani commented on includes/REST/ProductController.php:
```diff
@@ -0,0 +1,305 @@
+use WP_REST_Response;
+use WP_REST_Server;
+
+class ProductController extends WC_REST_Controller {
```
Pls extend the `WC_REST_Products_Controller`

mrabbani commented on includes/Bootstrap.php:
```diff
@@ -8,6 +8,7 @@
 namespace StorePulse\StoreGrowth;
 
 use StorePulse\StoreGrowth\Admin\AdminMenu;
+use StorePulse\StoreGrowth\Providers\ServiceProvider;
```
Pls do not register provider here

mrabbani commented on includes/Providers/ServiceProvider.php:
```diff
@@ -0,0 +1,35 @@
+use StorePulse\StoreGrowth\DependencyManagement\BootableServiceProvider;
+use StorePulse\StoreGrowth\REST\ProductController;
+
+class ServiceProvider extends BootableServiceProvider {
```
- Move the provider to `StorePulse\StoreGrowth\DependencyManagement\Providers` namespace
- Rename to `RestServiceProvider`
- Register to https://github.com/getdokan/storegrowth-sales-booster/blob/eddd2ffa6f973855abd6d4d90fbdb89db3c1915b/includes/DependencyManagement/Providers/ServiceProvider.php#L30

mrabbani commented on modules/bogo/assets/src/components/BasicInfo.jsx:
```diff
@@ -190,6 +109,47 @@ const BasicInfo = ({ clearErrors }) => {
     { key: 'categories', disabled: hidePremiumFeature, needUpgrade: hidePremiumFeature, value: __('Categories', 'storegrowth-sales-booster') },
   ];
 
+  function debounce(func, delay) {
```
Extract to a component to reuse in future.

kzamanbd committed:
Refactor product selection to use async select

Introduces ProductAsyncSelect component for improved product search and selection in BOGO settings. Refactors REST ProductController to extend WC_REST_Products_Controller and moves service provider registration to CommonServiceProvider. Updates Dokan integration to filter product queries by author for non-admins and cleans up related frontend logic.

mrabbani commented on integrations/includes/Dokan/Dashboard/Bogo.php:
```diff
@@ -37,8 +37,19 @@ private function init_hooks() {
+
+		$args['author'] = ! isset( $request['id'] ) ? dokan_get_current_user_id() : $request['id'];
+
+		return $args;
```
```php
if (  ! current_user_can(  'manage_woocommerce'  ) )  && function_exist(  'dokan_get_current_user' ) ) {
		$args[ 'author' ] = dokan_get_current_user_id();
}

return $args;
```

kzamanbd committed:
fix: Update product query args to restrict author assignment based on user capabilities

mrabbani committed:
Merge branch 'develop' into fix/bogo-handle-large-product

kzamanbd committed:
Merge branch 'develop' into fix/bogo-handle-large-product

kzamanbd committed:
Merge branch 'develop' into fix/bogo-handle-large-product

kzamanbd committed:
Refactor design value retrieval to Helper class

kzamanbd committed:
Merge branch 'develop' into fix/bogo-handle-large-product
